src/ServiceControl/MessageRedirects/Api/MessageRedirectsModule.cs

```diff
@@ -0,0 +1,202 @@
+
+                    var messageRedirect = redirects[messageRedirectId];
+
+                    if (messageRedirect == null)
+                    {
+                        return HttpStatusCode.NotFound;
+                    }
+
+                    var toMessageRedirectId = DeterministicGuid.MakeId(request.tophysicaladdress.ToLowerInvariant());
+
+                    if (redirects[toMessageRedirectId] != null)
+                    {
+                        return HttpStatusCode.Conflict;
+                    }
+
+                    var messageRedirectChanged = new MessageRedirectChanged
+                    {
+                        MessageRedirectId = messageRedirectId,
+                        PreviousToPhysicalAddress = messageRedirect.ToPhysicalAddress,
+                        FromPhysicalAddress = messageRedirect.FromPhysicalAddress,
+                        ToPhysicalAddress = messageRedirect.ToPhysicalAddress = request.tophysicaladdress.ToLowerInvariant(),
+                    };
+
+                    messageRedirect.LastModifiedTicks = DateTime.UtcNow.Ticks;
+
+                    redirects.Save(session);
+
+                    Bus.Publish(messageRedirectChanged);
+
+                    return HttpStatusCode.OK;
+                }
+            };
+
+            Delete["/redirects/{messageredirectid:guid}/"] = parameters =>
+            {
+                Guid messageRedirectId = parameters.messageredirectid;
+
+                using (var session = Store.OpenSession())
+                {
+                    var redirects = MessageRedirectsCollection.GetOrCreate(session);
+
+                    var messageRedirect = redirects[messageRedirectId];
+
+                    if (messageRedirect == null)
+                    {
+                        return HttpStatusCode.OK;
+                    }
+
+                    redirects.Redirects.Remove(messageRedirect);
+
+                    redirects.Save(session);
+
+                    Bus.Publish<MessageRedirectRemoved>(evt =>
+                    {
+                        evt.MessageRedirectId = messageRedirectId;
+                        evt.FromPhysicalAddress = messageRedirect.FromPhysicalAddress;
+                        evt.ToPhysicalAddress = messageRedirect.ToPhysicalAddress;
+                    });
+                }
+
+                return HttpStatusCode.OK;
+            };
+
+            Head["/redirects"] = _ =>
+            {
+                using (var session = Store.OpenSession())
+                {
+                    var redirects = MessageRedirectsCollection.GetOrCreate(session);
+
+                    return Negotiate
+                        .WithEtagAndLastModified(redirects.ETag, redirects.Redirects.Max(r => new DateTime(r.LastModifiedTicks)))
+                        .WithTotalCount(redirects.Redirects.Count);
+                }
+            };
+
+            Get["/redirects"] = _ =>
```

Comment:
Addressed in https://github.com/Particular/ServiceControl/pull/762/commits/9cbcb86519b7107bf92a5e7daf2fdb6f3f6c144c Used document last modified metadata instead of internal data which may not be complete or change when the document is updated